interop/src/client.rs

```diff
@@ -0,0 +1,242 @@
+// limitations under the License.
+
+
+use std::thread;
+use std::time::Duration;
+
+use grpc::{self, CallOption, Channel, RpcStatusCode, WriteFlags};
+use futures::{Future, Sink, Stream, future, stream};
+
+use grpc_proto::testing::test_grpc::{TestServiceClient, UnimplementedServiceClient};
+use grpc_proto::testing::empty::Empty;
+use grpc_proto::testing::messages::{EchoStatus, SimpleRequest, StreamingInputCallRequest,
+                                    StreamingOutputCallRequest};
+use grpc_proto::util;
+
+pub struct Client {
+    channel: Channel,
+    client: TestServiceClient,
+}
+
+impl Client {
+    pub fn new(ch: Channel) -> Client {
+        Client {
+            channel: ch.clone(),
+            client: TestServiceClient::new(ch),
+        }
+    }
+
+    pub fn empty_unary(&self) {
+        print!("testing empty unary ... ");
+        let req = Empty::new();
+        let resp = self.client.empty_call(req.clone()).unwrap();
+        assert_eq!(req, resp);
+        println!("pass");
+    }
+
+    pub fn large_unary(&self) {
+        print!("testing large unary ... ");
+        let mut req = SimpleRequest::new();
+        req.set_response_size(314159);
+        req.set_payload(util::new_payload(271828));
+        let resp = self.client.unary_call(req).unwrap();
+        assert_eq!(314159, resp.get_payload().get_body().len());
+        println!("pass");
+    }
+
+    pub fn client_streaming(&self) {
+        print!("testing client streaming ... ");
+        let reqs: Vec<grpc::Result<_>> = vec![27182, 8, 1828, 45904]
+            .into_iter()
+            .map(|s| {
+                let mut req = StreamingInputCallRequest::new();
+                req.set_payload(util::new_payload(s));
+                Ok((req, WriteFlags::default()))
+            })
+            .collect();
+        let (sender, receiver) = self.client.streaming_input_call();
+        sender.send_all(stream::iter(reqs)).wait().unwrap().0;
+        let resp = receiver.wait().unwrap();
+        assert_eq!(74922, resp.get_aggregated_payload_size());
+        println!("pass");
+    }
+
+    pub fn server_streaming(&self) {
+        print!("testing server streaming ... ");
+        let mut req = StreamingOutputCallRequest::new();
+        let sizes = vec![31415, 9, 2653, 58979];
+        for size in &sizes {
+            req.mut_response_parameters()
+                .push(util::new_parameters(*size));
+        }
+        let resp = self.client.streaming_output_call(req);
+        let resp_sizes = resp.map(|r| r.get_payload().get_body().len() as i32)
+            .collect()
+            .wait()
+            .unwrap();
+        assert_eq!(resp_sizes, sizes);
+        println!("pass");
+    }
+
+    pub fn ping_pong(&self) {
+        print!("testing ping pong ... ");
+        let (mut sender, mut receiver) = self.client.full_duplex_call();
+        let cases = vec![(31415, 27182), (9, 8), (2653, 1828), (58979, 45904)];
+        for (resp_size, payload_size) in cases {
```

Comment:
L102 only waits the stream, it does not check if the stream has exactly four responses.